examples/mixture/plot_gmm_selection.py

```diff
@@ -3,108 +3,161 @@
 Gaussian Mixture Model Selection
 ================================
 
-This example shows that model selection can be performed with
-Gaussian Mixture Models using :ref:`information-theoretic criteria (BIC) <aic_bic>`.
-Model selection concerns both the covariance type
-and the number of components in the model.
-In that case, AIC also provides the right result (not shown to save time),
-but BIC is better suited if the problem is to identify the right model.
-Unlike Bayesian procedures, such inferences are prior-free.
+This example shows that model selection can be performed with Gaussian Mixture
+Models (GMM) using :ref:`information-theory criteria <aic_bic>`. Model selection
+concerns both the covariance type and the number of components in the model.
 
-In that case, the model with 2 components and full covariance
-(which corresponds to the true generative model) is selected.
+In this case, both the Akaike Information Criterion (AIC) and the Bayes
+Information Criterion (BIC) provide the right result, but we only demo the
+latter as BIC is better suited to identify the true model among a set of
+candidates. Unlike Bayesian procedures, such inferences are prior-free.
 
 """
 
+# %%
+# Data generation
+# ---------------
+#
+# We generate two components (each one containing `n_samples`) by randomly
+# sampling the standard normal distribution as returned by `numpy.random.randn`.
+# One component is kept spherical yet shifted and re-scaled. The other one is
+# deformed to have a more general covariance matrix.
+
 import numpy as np
-import itertools
 
-from scipy import linalg
+n_samples = 500
+np.random.seed(0)
+C = np.array([[0.0, -0.1], [1.7, 0.4]])
+component_1 = np.dot(np.random.randn(n_samples, 2), C)  # general
+component_2 = 0.7 * np.random.randn(n_samples, 2) + np.array([-4, 1])  # spherical
+
+X = np.concatenate([component_1, component_2])
+
+# %%
+# We can visualize the different components:
+
 import matplotlib.pyplot as plt
-import matplotlib as mpl
 
-from sklearn import mixture
+plt.scatter(component_1[:, 0], component_1[:, 1], 0.8)
+plt.scatter(component_2[:, 0], component_2[:, 1], 0.8)
+plt.title("Gaussian Mixture components")
+plt.axis("equal")
+plt.show()
 
-# Number of samples per component
-n_samples = 500
+# %%
+# Model training and selection
+# ----------------------------
+#
+# We vary the number of components from 1 to 6 and the type of covariance
+# parameters to use:
+#
+# - `"full"`: each component has its own general covariance matrix.
+# - `"tied"`: all components share the same general covariance matrix.
+# - `"diag"`: each component has its own diagonal covariance matrix.
+# - `"spherical"`: each component has its own single variance.
+#
+# We score the different models and keep the best model (the lowest BIC).
```

Comment:
Since we modify the example, I think that we can add a bit more details.

```
The different models are evaluated by cross-validation using
:class:`~sklearn.model_selection.GridSearchCV` and a user-defined score
the function corresponding to the negative BIC score. The reason to take
a negative BIC score is due to the fact that
:class:`~sklearn.model_selection.GridSearchCV` is always **maximizing**
a score (i.e. higher is better).

The best set of parameters and estimator are stored in `best_parameters_`
and `best_estimator_`, respectively.
```